Multiple issues from OpenNMT-py

This AssertionError always happens when I try to torchdynamo.optimize transformer.

Hi @jansel Thanks for the fix. After applying your patch, we meet another issue now.
```
ERROR FROM offset=292 filename /home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/beam_search.py 284 AssertionError
========== TorchDynamo Stack Trace ==========
Traceback (most recent call last):
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/convert_frame.py", line 158, in _convert_frame_assert
    code = transform_code_object(frame.f_code, transform)
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/bytecode_transformation.py", line 284, in transform_code_object
    transformations(instructions, code_options)
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/convert_frame.py", line 134, in transform
    tracer.run()
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/symbolic_convert.py", line 274, in run
    and self.step()
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/symbolic_convert.py", line 252, in step
    getattr(self, inst.opname)(inst)
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/symbolic_convert.py", line 140, in wrapper
    return inner_fn(self, inst)
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/symbolic_convert.py", line 97, in impl
    self.push(fn_var.call_function(self, self.popn(nargs), {}))
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/variables/builtin.py", line 189, in call_function
    return variables.TensorVariable.create(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/variables/tensor.py", line 100, in create
    assert (
AssertionError: namedtuple?
========== Exception (above) while processing ==========
  File "onmt/bin/translate.py", line 52, in <module>
    main()
  File "onmt/bin/translate.py", line 48, in main
    translate(opt)
  File "onmt/bin/translate.py", line 25, in translate
    translator.translate(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 400, in translate
    batch_data = self.translate_batch(batch, data.src_vocabs, attn_debug)
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 589, in translate_batch
    return self._translate_batch_with_strategy(batch, src_vocabs,
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 665, in _translate_batch_with_strategy
    def _translate_batch_with_strategy(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 695, in _translate_batch_with_strategy
    "gold_score": self._gold_score(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 665, in _translate_batch_with_strategy
    def _translate_batch_with_strategy(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 665, in _translate_batch_with_strategy
    def _translate_batch_with_strategy(
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py", line 665, in _translate_batch_with_strategy
    def _translate_batch_with_strategy(
  [Previous line repeated 101 more times]
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/beam_search.py", line 260, in update_finished
    def update_finished(self):
  File "/home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/beam_search.py", line 260, in update_finished
    def update_finished(self):
========== End debug info ==========
--------------finish one step----------
TypeError: <lambda>() missing 1 required positional argument: 'select_indices'
ERROR RUNNING GUARDS _translate_batch_with_strategy /home/lesliefang/pytorch_1_7_1/TorchDynamo/opennmt/opennmt-py/onmt/translate/translator.py:665
  ___guarded_code.valid and
  ___check_type_id(results, 93934403079104) and
  results.keys() == {'predictions', 'batch', 'attention', 'scores', 'gold_score'} and
  ___check_type_id(___stack0, 93934403067808) and
  ___tuple_iterator_len(___stack0) == 174 and
  ___check_type_id(___stack1, 93934403067392) and
  len(___stack1) == 2 and
  ___check_type_id(src_vocabs, 93934403080768) and
  len(src_vocabs) == 0 and
  ___check_type_id(decode_strategy, 93934469735584) and
  ___check_tensors(memory_bank, ___stack1[0], ___stack1[1], decoder_input, memory_lengths, select_indices)
NULL ERROR: torchdynamo/_eval_frame.c:190
run.sh: line 13: 15577 Aborted                 (core dumped) python onmt/bin/translate.py --model checkpoint_step_250000.pt --src valid_part.en --shard_size 1000000 --output valid.en.out --beam_size 4 --max_length 200 --length_penalty "wu" --alpha 0.6 --coverage_penalty "wu" --beta 0.6 --batch_size 1 --report_time --fp32

```

Here is some debug information at `/home/lesliefang/pytorch_1_7_1/TorchDynamo/torchdynamo/torchdynamo/variables/tensor.py", line 100, in create
    assert (
AssertionError: namedtuple?`
```
(Pdb) p example_value.__class__.__module__
'torch.fx.immutable_collections'
(Pdb) p hasattr(example_value, "_fields")
False
(Pdb) p example_value
[]
(Pdb) p isinstance(example_value, (tuple, list))
True
(Pdb) p istype(example_value, (tuple, list))
False
```
Not sure what's the right way to fix it :)?

--- a/tests/test_repros.py
+++ b/tests/test_repros.py
@@ -925,3 +925,29 @@ class ReproTests(torchdynamo.testing.TestCase):
             )
         self.assertEqual(cnt.frame_count, 1)
         self.assertEqual(cnt.op_count, ifdyn(28, 14))
+
+    def test_recursive_map(self):
+        # https://github.com/facebookresearch/torchdynamo/issues/132
+        def _recursive_map(struct, batch_dim=0):
+            for k, v in struct.items():
+                if v is not None:
+                    if isinstance(v, dict):
+                        _recursive_map(v)
+                    else:
+                        struct[k] = v
+
+        def toy_example(a, b, v):
+            x = a / (torch.abs(a) + 1)
+            if v is not None:
+                _recursive_map(v)
+            return x * b
+
+        cnt = torchdynamo.testing.CompileCounter()
+        with torchdynamo.optimize(cnt):
+            toy_example(
+                torch.randn(10),
+                torch.randn(10),
+                {"layer0": {"memory_keys": torch.randn(10)}},
+            )
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(cnt.op_count, 4)

--- a/torchdynamo/symbolic_convert.py
+++ b/torchdynamo/symbolic_convert.py
@@ -501,6 +501,7 @@ class InstructionTranslatorBase(object):
                     BaseListVariable,
                     UserDefinedVariable,
                     BaseUserFunctionVariable,
+                    ConstDictVariable,
                 ),
             )
             and isinstance(right, ConstantVariable)
@@ -1031,7 +1032,9 @@ class InstructionTranslator(InstructionTranslatorBase):
 
         # TODO(jansel): figure out why the following is needed for detectron2_maskrcnn
         for val in self.symbolic_locals.values():
-            if isinstance(val, (ListIteratorVariable, BaseListVariable)):
+            if isinstance(
+                val, (ListIteratorVariable, BaseListVariable, ConstDictVariable)
+            ):
                 self.output.guards.update(val.guards)
 
         self._freevars_ids = dict()
@@ -1160,6 +1163,7 @@ class InliningInstructionTranslator(InstructionTranslatorBase):
             assert tracer.symbolic_result.as_python_constant() is None
             return ListIteratorVariable(
                 tracer.generated_items,
+                mutable_local=MutableLocal(),
                 **VariableTracker.propagate(tracer.symbolic_result),
             )
         else:
@@ -1188,8 +1192,6 @@ class InliningInstructionTranslator(InstructionTranslatorBase):
         )
         self.symbolic_result = None
         self.closure_cells = closure_cells
-        # self.funcvar = funcvar
-        # self.parent = parent
 
     def STORE_DEREF(self, inst):
         if inst.argval in self.closure_cells:

